### PR TITLE
Adds Parity PoA chain mode.

### PR DESCRIPTION
**Pull request by MicahZoltu**
Using PoA chain requires changing `start.sh` to point at the other toml file.  This can be improved in the future via a smarter scirpt and docker COMMAND arguments.

PoA has 1 second block time, which allows testing of block timing delays.

Also did some cleanup/minor quality of life improvements to contract deployer library.

### CODE REVIEW

File: source/libraries/ContractDeployer.ts [hunk 1]
```diff
@@ -3,13 +3,14 @@
...
-import * as EthjsQuery from "ethjs-query";
+import EthjsAbi = require("ethjs-abi");
+import EthjsContract = require("ethjs-contract");
+import EthjsQuery = require("ethjs-query");
+import { TransactionReceipt } from 'ethjs-shared';
 // TODO: Update TS type definition for ContractBlockchainData to allow for empty object (e.g. upload() & uploadAndAddToController())?
 import { ContractBlockchainData, ContractReceipt } from "contract-deployment";
 import { Contract, parseAbiIntoMethods } from "./AbiParser";
-import { generateTestAccounts, padAndHexlify, stringTo32ByteHex, waitForTransactionToBeSealed } from "./HelperFunctions";
+import { generateTestAccounts, padAndHexlify, stringTo32ByteHex, waitForTransactionReceipt, sleep } from "./HelperFunctions";
```
Comment (adrake33) [inline source/libraries/ContractDeployer.ts]: Sleep can be removed here.

File: source/libraries/ContractDeployer.ts [hunk 2]
```diff
@@ -140,64 +143,69 @@ export class ContractDeployer {
...
                     // this.contracts[contractName] = this.applySignature(contractName, this.contracts[contractName].address)
                 } else {
-                    uploadedContractPromises.push(this.uploadAndAddToController(contractFileName));
+                    promises.push(this.uploadAndAddToController(contractFileName));
                 }
             }
         }
 
-        await Promise.all(uploadedContractPromises);
+        const latestBlock = await this.ethjsQuery.getBlockByNumber('latest', true);
```
Comment (adrake33) [inline source/libraries/ContractDeployer.ts]: Why did you add this line?

File: source/libraries/ContractDeployer.ts [hunk 3]
```diff
@@ -249,7 +256,6 @@ export class ContractDeployer {
             throw new Error("Unable to get address for new categorical market.");
         }
         const createMarketHash = await targetReportingWindow.createMarket.bind({ value: marketCreationFee })(endTime, numOutcomes, numTicks, feePerEthInWei, denominationToken, designatedReporter);
-        waitForTransactionToBeSealed(this.ethjsQuery, createMarketHash);
```
Comment (adrake33) [inline source/libraries/ContractDeployer.ts]: We don't need to wait before calling `market.getTypeName`?

File: source/libraries/HelperFunctions.ts [hunk 4]
```diff
@@ -74,14 +75,15 @@ export async function initializeTestRpcClientOptions(ethHttpProviderPort: number
     return { gasLimit: GAS_BLOCK_AMOUNT, accounts: accountOptions };
 }
 
-async function sleep(milliseconds: number): Promise<object> {
-    return new Promise((resolve, reject) => setTimeout(resolve, milliseconds));
+export async function sleep(milliseconds: number): Promise<void> {
```
Comment (adrake33) [inline source/libraries/HelperFunctions.ts]: Do we actually need to export this?  I didn't see where it's being used outside of this file.

File: typings/ethjs-shared.d.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        logIndex: BN;
+        transactionIndex: BN;
+        transactionHash: string;
+        blockHash: string;
+        blockNumber: BN;
+        address: string;
+        data: string;
+        topics: Array<string>;
+    }
+}
```
Comment (adrake33) [inline typings/ethjs-shared.d.ts]: Nice, thanks for adding these type definition files!

File: parity-dev-node/aura-chain-spec.json [hunk 6]
```diff
@@ -0,0 +1,34 @@
+{
+    "name": "AuraChain",
```
[the remark this answers is not in the document]
Comment (MicahZoltu) [reply parity-dev-node/aura-chain-spec.json]: Aura is one of the Parity Proof of Authority algorithms, like Clique is for Geth.  Parity also has Tendermint Proof of Authority I believe.

File: parity-dev-node/start.sh [hunk 7]
```diff
@@ -4,16 +4,9 @@
...
 # launch parity in the foreground
-/parity/parity \
-	--chain dev \
-	--gasprice 2 \
-	--no-discovery \
-	--force-ui --ui-no-validation --ui-interface 0.0.0.0 \
-	--jsonrpc-interface all --jsonrpc-cors "*" --jsonrpc-hosts all --jsonrpc-apis web3,eth,net,personal,parity,parity_set,traces,rpc,parity_accounts \
-	--unlock 0x00a329c0648769a73afac7f9381e08fb43dbea72 --password /parity/password \
-	&
+/parity/parity --config /parity/instant-seal-config.toml --gasprice 1 &
```
[the remark this answers is not in the document]
Comment (MicahZoltu) [reply parity-dev-node/start.sh]: The trailing ampersand is necessary to launch Parity in the background, though your comment makes me realize that the comment above this line is wrong.  :)

File: parity-dev-node/aura-chain-spec.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (adrake33) [reply parity-dev-node/aura-chain-spec.json]: Ah, that's why I hadn't seen it before.  Ok, good to know.

File: parity-dev-node/start.sh [hunk 8]
```diff
@@ -4,16 +4,9 @@
...
 
 # spin until node is connectable
 while ! curl --silent --show-error localhost:8545 -H "Content-Type: application/json" -X POST --data '{"jsonrpc":"2.0","method":"net_version","id": 1}'; do sleep 0.1; done
-curl --silent --show-error localhost:8545 -H "Content-Type: application/json" -X POST --data '{"jsonrpc":"2.0","method":"eth_sendTransaction","params":[{"value":"0x0","to":"0x0000000000000000000000000000000000000000","from":"0x00a329c0648769a73afac7f9381e08fb43dbea72","data":"0x","gasPrice":"0x3B9ACA00"}], "id": 1}'
+curl --silent --show-error localhost:8545 -H "Content-Type: application/json" -X POST --data '{"jsonrpc":"2.0","method":"eth_sendTransaction","params":[{"value":"0x0","to":"0x0000000000000000000000000000000000000000","from":"0x913da4198e6be1d5f5e4a40d0667f70c0b5430eb","data":"0x","gasPrice":"0x3B9ACA00"}], "id": 1}'
 fg
```
[the remark this answers is not in the document]
Comment (MicahZoltu) [reply parity-dev-node/start.sh]: We want Parity to eventually run in the foreground, but we also want to run the while loop and then send that transaction once it starts up.  To achieve that, we start Parity in the background, then run the curls that we need, then when they are done we bring Parity back to the foreground.

File: source/deployment/deployContracts.ts [hunk 9]
```diff
@@ -3,8 +3,8 @@
 import * as fs from "async-file";
 import * as getPort from "get-port";
 import * as path from "path";
-import * as EthjsHttpProvider from "ethjs-provider-http";
-import * as EthjsQuery from "ethjs-query";
+import EthjsHttpProvider = require('ethjs-provider-http');
+import EthjsQuery = require("ethjs-query");
```
[the remark this answers is not in the document]
Comment (MicahZoltu) [reply source/deployment/deployContracts.ts]: It turns out `import * as EthjsQuery from 'ethjs-query';` is not correct and only works by happenstance and will likely break in the future.  Fixing these up was just some code cleanup.  Note, this style is only necessary for imports that you need to `new` directly.  For an import that is just a property bag, the other style is still appropriate (as far as I understand).

File: source/libraries/AbiParser.ts [hunk 10]
```diff
@@ -1,7 +1,7 @@
 import * as EthjsAbi from 'ethjs-abi';
 import * as EthjsQuery from 'ethjs-query';
 import { CompilerOutputContractAbiFunction, CompilerOutputContractAbiEvent } from 'solc';
-import { waitForTransactionToBeSealed } from './HelperFunctions';
+import { waitForTransactionReceipt } from './HelperFunctions';
```
[the remark this answers is not in the document]
Comment (MicahZoltu) [reply source/libraries/AbiParser.ts]: Fixed.

File: source/libraries/ContractDeployer.ts (hunk 1 above)
Comment (MicahZoltu) [reply source/libraries/ContractDeployer.ts]: Fixed.

File: source/libraries/ContractDeployer.ts (hunk 2 above)
Comment (MicahZoltu) [reply source/libraries/ContractDeployer.ts]: Oops, debugging.  Will remove.

File: parity-dev-node/start.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (adrake33) [reply parity-dev-node/start.sh]: Oops, yeah, I didn't catch that the ampersand was also there before the change.  Never mind.

File: source/libraries/ContractDeployer.ts (hunk 3 above)
Comment (MicahZoltu) [reply source/libraries/ContractDeployer.ts]: This call was missing the `await` so it wasn't actually waiting.  This line was effectively a no-op, so I removed it.  We _may_ need to wait here, and I'll be digging into the market creation issue tomorrow to find out.  I removed this because at the moment, it wasn't doing anything but clogging up the intertubes.  😄

File: parity-dev-node/start.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (adrake33) [reply parity-dev-node/start.sh]: Ah, I see now.  Ok, that makes more sense.

File: source/libraries/HelperFunctions.ts (hunk 4 above)
Comment (MicahZoltu) [reply source/libraries/HelperFunctions.ts]: In the past I have found that having an async sleep can be really useful when debugging, so even though we may only use it directly here, having it exported can be really useful.  Since this is in a `HelperFunctions` file I made an exception to my usual personal rule of "only export things you actually need to export", since it is so useful for debugging.

File: source/deployment/deployContracts.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (adrake33) [reply source/deployment/deployContracts.ts]: That's weird, why does it actually work if it's not correct?

File: source/libraries/ContractDeployer.ts (hunk 3 above)
Comment (adrake33) [reply source/libraries/ContractDeployer.ts]: Haha, oops, good catch!  I was probably sleep-deprived when I added that line. :P

File: source/deployment/deployContracts.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (MicahZoltu) [reply source/deployment/deployContracts.ts]: Apparently just because of the way TypeScript compiler happens to be implemented makes it so it happens to be valid and they haven't bother to write the code that will trigger an error when someone does this incorrectly.  This Q&A talks a bit about the situation: https://stackoverflow.com/questions/39415661/what-does-resolves-to-a-non-module-entity-and-cannot-be-imported-using-this

File: source/libraries/HelperFunctions.ts [hunk 11]
```diff
@@ -74,14 +75,15 @@ export async function initializeTestRpcClientOptions(ethHttpProviderPort: number
...
 
-export async function waitForTransactionToBeSealed(ethjsQuery: EthjsQuery, transactionHash: string): Promise<void> {
-    let transaction = await ethjsQuery.getTransactionByHash(transactionHash);
-    while (transaction.blockNumber == null) {
+export async function waitForTransactionReceipt(ethjsQuery: EthjsQuery, transactionHash: string): Promise<TransactionReceipt> {
+    let transactionReceipt = await ethjsQuery.getTransactionReceipt(transactionHash);
+    while (!transactionReceipt) {
         await sleep(ETHEREUM_POLLING_INTERVAL_MILLISECONDS);
-        transaction = await ethjsQuery.getTransactionByHash(transactionHash);
+        transactionReceipt = await ethjsQuery.getTransactionReceipt(transactionHash);
```
Comment (epheph) [inline source/libraries/HelperFunctions.ts]: Is what we really want to wait the transaction receipt, or for it to be mined? Since we've seen parity give receipts for un-mined TX. The contract_address will be correct, but an eth_call() won't have the data available to it yet

Comment (MicahZoltu) [reply source/libraries/HelperFunctions.ts]: Ah, good catch.  I meant to wait for the transaction receipt to have a block.  Will fix in next PR.
